[processing] Port some stuff to c++

nyalldawson opened the pull request:
This ports parts of the provider and alglist processing classes to c++. As per the commit logs:

c++ base class for providers
-------------------------------------
Nothing particularly exciting here yet, but this commit moves the definition of the provider base class to a c++ QgsProcessingProvider abstract base class.

As part of this some existing python methods were renamed to make their use clearer and to fit with the QGIS c++ api conventions:

- getName was renamed to id
- getDescription was renamed to name
- getIcon was renamed to icon

New c++ QgsProcessingRegistry class
----------------------------------------------------
Intended as a registry for the various processing components, including providers, algorithms and various parameters and outputs.

Currently handles only registration of providers, as a step toward removing processing's algList (this requires first porting the algorithm class to c++)

A QgsProcessingRegistry instance is attached to QgsApplication, and is retrieved by QgsApplication::processingRegistry()

nyalldawson committed:
[processing] c++ base class for providers

Nothing particularly exciting here yet, but this commit
moves the definition of the provider base class to a c++
QgsProcessingProvider abstract base class.

As part of this some existing python methods were renamed
to make their use clearer and to fit with the QGIS c++
api conventions:

- getName was renamed to id
- getDescription was renamed to name
- getIcon was renamed to icon

These API breaks are documented

nyalldawson committed:
[processing] New c++ QgsProcessingRegistry class

Intended as a registry for the various processing components,
including providers, algorithms and various parameters and outputs.

Currently handles only registration of providers, as a step toward
removing processing's algList (this requires first porting
the algorithm class to c++)

A QgsProcessingRegistry instance is attached to QgsApplication,
and is retrieved by QgsApplication::processingRegistry()

nyalldawson committed:
Add some unit tests for QgsProcessingRegistry

nirvn commented:
@nyalldawson , FYI, when updating the processing's modeler to draw vectors instead of bitmaps, I noticed we'll need a way for providers and algorithms to return the file path of the icon (to be able to render .SVGs to a QPicture object).

nyalldawson committed:
[processing] Allow providers to return path to SVG icon

nyalldawson committed:
[processing] Move some SVG icons out of plugin to themes

...and swap to using QgsApplication methods for retrieving cached icons,
implement svgIconPath for providers with SVG icons

nyalldawson commented:
> FYI, when updating the processing's modeler to draw vectors instead of bitmaps, I noticed we'll need a way for providers and algorithms to return the file path of the icon (to be able to render .SVGs to a QPicture object).

@nirvn done, for those providers with a svg icon. Missing are saga, otb and lastools. The existing provider svgs could do with some tweaking too IMO to better work at the small icon sizes (eg gdal/grass icons are very busy)

nirvn commented:
@nyalldawson , 👍 . The same svgIconPath() function will be needed when porting Algorithm to QgsProcessingAlgorithm.

You are right re existing SVGs in need of tweaking. That's on my to-do list.

alexbruy commented:
Cool stuff.

DelazJ commented on python/core/processing/qgsprocessingprovider.sip:
```diff
@@ -0,0 +1,60 @@
+    virtual QString name() const = 0;
+
+    /**
+     * Returns true if the provider can be actived, or false if it is cannot be actived (eg due to
```
activ**at**ed
it ~~is~~ cannot
eg or e.g. ? (let us be en-US :-) )

nyalldawson commented:
@nirvn ok, over to you for visual improvements ;)
